src/FormField/MultiLine.php

```diff
@@ -0,0 +1,764 @@
+    {
+        if (!is_callable($fx)) {
+            throw new Exception('Function is required for onChange event.');
+        }
+        if ($fields) {
+            $this->eventFields = $fields;
+        }
+
+        $this->changeCb = $fx;
+    }
+
+    /**
+     * Input field collecting multiple rows of data.
+     *
+     * @throws \atk4\core\Exception
+     *
+     * @return string
+     */
+    public function getInput()
+    {
+        return $this->app->getTag('input', [
+            'name'        => $this->short_name,
+            'type'        => 'hidden',
+            'value'       => $this->getValue(),
+            'readonly'    => true,
+        ]);
+    }
+
+    /**
+     * Get multiline initial field value.
+     * Value is based on model set and will
+     * output data rows as json string value.
+     *
+     *
+     * @throws \atk4\core\Exception
+     *
+     * @return false|string
+     */
+    public function getValue()
+    {
+        $m = null;
+        $data = [];
+
+        //set model according to model reference if set; or simply the model pass to it.
+        if ($this->model->loaded() && $this->modelRef) {
+            $m = $this->model->ref($this->modelRef);
+        } elseif (!$this->modelRef) {
+            $m = $this->model;
+        }
+        if ($m) {
+            foreach ($m as $id => $row) {
+                $d_row = [];
+                foreach ($this->rowFields as $fieldName) {
+                    $field = $m->getElement($fieldName);
+                    if ($field->isEditable()) {
+                        $value = $row->get($field);
+                    } else {
+                        $value = $this->app->ui_persistence->_typecastSaveField($field, $row->get($field));
+                    }
+                    $d_row[$fieldName] = $value;
+                }
+                $data[] = $d_row;
+            }
+        }
+
+        return json_encode($data, JSON_UNESCAPED_UNICODE);
+    }
+
+    /**
+     * Validate each row and return errors if found.
+     *
+     * @param $rows
+     *
+     * @throws \atk4\core\Exception
+     *
+     * @return array|null
+     */
+    public function validate($rows)
+    {
+        $rowErrors = [];
+        $m = $this->getModel();
+
+        foreach ($rows as $row => $cols) {
+            $rowId = $this->getMlRowId($cols);
+            foreach ($cols as $col) {
+                $fieldName = key($col);
+                if ($fieldName === '__atkml' || $fieldName === $m->id_field) {
+                    continue;
+                }
+                $value = $col[$fieldName];
+
+                try {
+                    $field = $m->getElement($fieldName);
+                    // save field value only if field was editable in form at all
+                    if (!$field->read_only) {
+                        $m[$fieldName] = $this->app->ui_persistence->typecastLoadField($field, $value);
+                    }
+                } catch (\atk4\core\Exception $e) {
```

Comment:
we should only catch and display validation errors (they have a dedicated class).

Reply:
Was not sure how to check for post load validation error. Thought that typecastLoadField was supposed to do so. Let me know how this should work and I can change it.